### DISCUSSION TITLE
Reloading x-for throws " Cannot read property 'remove' of undefined"

### DISCUSSION THREAD

**Discussion by yabdab** [alpinejs/alpine#1623] [7. Bugs (archived)]
Reloading a x-for template results in " Cannot read property 'remove' of undefined" error.

See example here: https://jsfiddle.net/yabdab/esvjwLtg/19/

Click first time works fine, click second time to reload and it fails.

**Comment by atompulse** [comment]
+1 for this issue

**Comment by danddanddand** [comment]
I think the error is just that you're x-modeling to var that doesn't exist. I turned on the experimental debug mode (https://github.com/alpinejs/alpine/pull/1607) and it shows elements with error so you can see it better.

https://jsfiddle.net/jh5nax9u/

**Reply by yabdab** [reply]
Not sure I follow? The options var is hardcoded into the script. Can you elaborate?

**Reply by danddanddand** [reply]
This line, shippingType is not defined in your test() func:

`<input type="radio" name="shippingType" x-model="shippingType" :value="option.value" @click="$refs.shippingType.classList.remove('error')">`

**Reply by yabdab** [reply]
Yep, yer right. I forgot to add that var when recreating the bug. That has nothing to do with the bug however.

Here is the revised fiddle with shippingType added... https://jsfiddle.net/yabdab/esvjwLtg/21/

**Reply by danddanddand** [reply]
Okay thanks, the other exception was getting in my way of seeing the real one. You're right that looks like bug for sure

**Reply by danddanddand** [reply]
I did some more digging and if you change your :key to be something more meaningful this will work as you expect it. For example, https://codepen.io/danddanddand/pen/Rwpzdby is boiled down version of your code ~~with index used as key~~ (update, don't do this, use string/number)

**Reply by danddanddand** [reply]
I don't fully understand the exception alpine is throwing, but it's for sure getting messed up in it's lookups I think because the key isn't unique. Still might be way for alpine to test for this better or error out better

**Reply by danddanddand** [reply]
Here is your example with the more meaningful index... The :key must be getting toString() of object or something weird. using option.value ~~probably would work~~ is recommended method (not index as in this sample). https://jsfiddle.net/rsv9uo5n/

**Reply by KevinBatdorf** [reply]
Keys need to be primitive types like numbers or strings. Same goes for other libraries too like React, etc. 

You should also "never" use the `index` either. Index will work if you're not rearranging, the items but both Alpine and React will add for you as a backup it if you don't include it. 

Adding it is an anti pattern as it adds no value and promotes bad habits (consider new users copy/pasting code, etc). Just a tip, not trying to sound negative

**Comment by calebporzio** [comment]
The issue is that you are using an entire object as the `:key` for the x-for.

Try: `<template x-for="option in options" :key="option.value" hidden>` (notice the added `.value`)

I will investigate and see if there is a fix for this case or a better error to throw - however, x-for keys should be something stringable, not objects.

If you think a piece of documentation would have warned you against this, please PR a note in the docs - thanks!

**Reply by calebporzio** [reply]
Ok, I added a warning for the case when the `:key` isn't stringable: https://github.com/alpinejs/alpine/pull/1639

It will be available in the next release and hopefully prevent others from tripping up on this. Thanks!

**Reply by yabdab** [reply]
Thanks Caleb. My code worked fine in v2 is why I never noticed it. 
Your explanation of what the key should be makes total sense.

Thanks again for all your hard work on Alpine. 🤘
